### PR TITLE
Use optional exchange of ghost data in DoFHandler::distribute_dofs()

### PR DESCRIPTION
**Pull request by bangerth**
Specifically, avoid it if possible.

This is a follow-up to #4855 that uses the functionality I introduced in #4858 where one can select whether or not one wants to do the ghost exchange. This also contains the changes promised in the review of #4855.

In reference to #3511 and #1366. Also #4704.

### CODE REVIEW

File: source/dofs/dof_handler_policy.cc [hunk 1]
```diff
@@ -3205,97 +3205,67 @@ namespace internal
...
-            // only happen if the complete data set had previously
-            // been sent, and in that case, (i) the user flag must have been
-            // deleted, and (ii) the current set of DoF indices must already
-            // be complete
+            // this function should only be called on ghost cells, and
+            // on top of that, only on cells that have not been
+            // completed -- which we indicate via the user flag.
+            // check both
+            Assert (cell->is_ghost(), ExcInternalError());
+            Assert (cell->user_flag_set(), ExcInternalError());
```
Comment (tjhei) [inline source/dofs/dof_handler_policy.cc]: Could it happen that we receive the same cell from two processors in the same round and we remove the flag after we unpack the first message? Let's talk about this if you have a minute.

Comment (bangerth) [inline source/dofs/dof_handler_policy.cc]: I'm receiving data on a ghost cell, for which there is exactly one, unique owner from which we must have gotten the information. So I don't think that can happen.

Comment (tjhei) [inline source/dofs/dof_handler_policy.cc]: Yes, of course.

**Comment by tamiko** [comment]
@bangerth I suspect that this, or one of the previous PRs also changed the following (debug) test output:
https://cdash.kyomu.43-1.org/viewTest.php?onlydelta&buildid=10197

Can you please verify and update the tests?

**Comment by bangerth** [comment]
Hm, that seems odd. I don't think this patch had anything to do with these failures. But I may be missing something -- how did you infer which patch may have been responsible for these failures?

**Pull request merged**